File: go/src/228.汇总区间.go
/*
 * @lc app=leetcode.cn id=228 lang=golang
 *
 * [228] 汇总区间
 */

// @lc code=start
func summaryRanges(nums []int) []string {
	n := len(nums)
	if n == 0 {
		return nil
	}

	res := []string{}
	for i := 0; i < n; i++ {
		j := i + 1
		for j < n && nums[j] == nums[j - 1] + 1 {
			j++
		}
		if j > i + 1 {
			res = append(res, strconv.Itoa(nums[i]) + "->" + strconv.Itoa(nums[j - 1]))
		} else {
			res = append(res, strconv.Itoa(nums[i]))
		}
		i = j - 1
	}

	return res
}
// @lc code=end

File: rust/src/228.汇总区间.rs
/*
 * @lc app=leetcode.cn id=228 lang=rust
 *
 * [228] 汇总区间
 */

// @lc code=start
impl Solution {
    pub fn summary_ranges(nums: Vec<i32>) -> Vec<String> {
        let n = nums.len();
        if n == 0 {
            return vec![];
        }

        let mut res = vec![];
        let mut start_index = 0;
        for i in 1..n {
            if nums[i] == nums[i - 1] + 1 {
                continue;
            }

            if i > start_index + 1 {
                res.push(format!("{}->{}", nums[start_index], nums[i - 1]));
            } else {
                res.push(format!("{}", nums[start_index]));
            }
            start_index = i;
        }

        if start_index < n - 1 {
            res.push(format!("{}->{}", nums[start_index], nums[n - 1]));
        } else {
            res.push(format!("{}", nums[start_index]));
        }

        res
    }
}
// @lc code=end
